pymc_bart/tree.py

```diff
@@ -52,164 +106,161 @@ class Tree:
         "output",
     )
 
-    def __init__(self, tree_structure, idx_leaf_nodes, output):
+    def __init__(
+        self,
+        tree_structure: Dict[int, Node],
+        idx_leaf_nodes: Optional[npt.NDArray[np.int_]],
+        output: npt.NDArray[np.float_],
+    ) -> None:
         self.tree_structure = tree_structure
         self.idx_leaf_nodes = idx_leaf_nodes
         self.output = output
 
     @classmethod
-    def new_tree(cls, leaf_node_value, idx_data_points, num_observations, shape):
+    def new_tree(
+        cls,
+        leaf_node_value: float,
+        idx_data_points: Optional[npt.NDArray[np.int_]],
+        num_observations: int,
+        shape: int,
+    ) -> "Tree":
         return cls(
             tree_structure={
-                0: Node.new_leaf_node(0, value=leaf_node_value, idx_data_points=idx_data_points)
+                0: Node.new_leaf_node(
+                    index=0, value=leaf_node_value, idx_data_points=idx_data_points
+                )
             },
-            idx_leaf_nodes=[0],
+            idx_leaf_nodes=np.array([0]),
             output=np.zeros((num_observations, shape)).astype(config.floatX).squeeze(),
         )
 
-    def __getitem__(self, index):
+    def __getitem__(self, index) -> Node:
         return self.get_node(index)
 
-    def __setitem__(self, index, node):
+    def __setitem__(self, index, node) -> None:
         self.set_node(index, node)
 
-    def copy(self):
-        tree = {
+    def copy(self) -> "Tree":
+        tree: Dict[int, Node] = {
             k: Node(v.index, v.value, v.idx_data_points, v.idx_split_variable)
             for k, v in self.tree_structure.items()
         }
-        return Tree(tree, self.idx_leaf_nodes.copy(), self.output.copy())
+        idx_leaf_nodes = self.idx_leaf_nodes.copy() if self.idx_leaf_nodes is not None else None
+        output = self.output.copy() if self.output is not None else None
+        return Tree(tree_structure=tree, idx_leaf_nodes=idx_leaf_nodes, output=output)
 
-    def get_node(self, index) -> "Node":
+    def get_node(self, index: int) -> Node:
         return self.tree_structure[index]
 
-    def set_node(self, index, node):
+    def set_node(self, index: int, node: Node) -> None:
         self.tree_structure[index] = node
-        if node.is_leaf_node():
-            self.idx_leaf_nodes.append(index)
+        if node.is_leaf_node() and self.idx_leaf_nodes is not None:
+            # self.idx_leaf_nodes.append(index)
+            self.idx_leaf_nodes = np.append(self.idx_leaf_nodes, index)
```

Comment:
```suggestion
            self.idx_leaf_nodes.append(index)
```

Reply:
np.append is very slow compared to the append method from list, we should avoid it as we will need to perform this operation thousands of times